File: Android/dokit/src/main/java/com/didichuxing/doraemonkit/kit/alignruler/AlignRulerLineDoKitView.java
package com.didichuxing.doraemonkit.kit.alignruler;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.WindowManager;
import android.widget.FrameLayout;

import com.didichuxing.doraemonkit.DoKit;
import com.didichuxing.doraemonkit.util.ActivityUtils;
import com.didichuxing.doraemonkit.util.ConvertUtils;
import com.didichuxing.doraemonkit.util.ScreenUtils;
import com.didichuxing.doraemonkit.R;
import com.didichuxing.doraemonkit.kit.core.AbsDoKitView;
import com.didichuxing.doraemonkit.kit.core.DoKitViewLayoutParams;

/**
 * Created by jintai on 2019/09/26.
 */

public class AlignRulerLineDoKitView extends AbsDoKitView implements AlignRulerMarkerDoKitView.OnAlignRulerMarkerPositionChangeListener {
    private AlignRulerMarkerDoKitView mMarker;
    private AlignLineView mAlignInfoView;

    @Override
    public void onCreate(Context context) {

    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mMarker.removePositionChangeListener(this);
    }

    @Override
    public View onCreateView(Context context, FrameLayout view) {
        return LayoutInflater.from(context).inflate(R.layout.dk_float_align_ruler_line, view, false);
    }


    @Override
    public void initDokitViewLayoutParams(DoKitViewLayoutParams params) {
        params.flags = WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE | WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE;
        params.height = DoKitViewLayoutParams.MATCH_PARENT;
        params.width = DoKitViewLayoutParams.MATCH_PARENT;
    }

    @Override
    public void onViewCreated(FrameLayout view) {
        postDelayed(new Runnable() {
            @Override
            public void run() {
                mMarker = DoKit.getDoKitView(ActivityUtils.getTopActivity(), AlignRulerMarkerDoKitView.class);
                if (mMarker != null) {
                    mMarker.addPositionChangeListener(AlignRulerLineDoKitView.this);
                }
            }
        }, 100);
        setDoKitViewNotResponseTouchEvent(getDoKitView());
        mAlignInfoView = findViewById(R.id.info_view);
    }

    @Override
    public void onPositionChanged(int x, int y) {
        /**
         * 限制边界
         */
        if (!isNormalMode()) {
            int iconSize = ConvertUtils.dp2px(30);
            if (y <= iconSize) {
                y = iconSize;
            }

            if (ScreenUtils.isPortrait()) {
                if (y >= getScreenLongSideLength() - iconSize) {
                    y = getScreenLongSideLength() - iconSize;
                }
            } else {
                if (y >= getScreenShortSideLength() - iconSize) {
                    y = getScreenShortSideLength() - iconSize;
                }
            }


            if (x <= iconSize) {
                x = iconSize;
            }
            if (ScreenUtils.isPortrait()) {
                if (x >= getScreenShortSideLength() - iconSize) {
                    x = getScreenShortSideLength() - iconSize;
                }
            } else {
                if (x >= getScreenLongSideLength() - iconSize) {
                    x = getScreenLongSideLength() - iconSize;
                }
            }
        }


        mAlignInfoView.showInfo(x, y);
    }

    @Override
    public boolean canDrag() {
        return false;
    }

    @Override
    public boolean restrictBorderline() {
        return true;
    }

    public AlignLineView getAlignInfoView() {
        return mAlignInfoView;
    }
}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/alignruler/AlignRulerLineDokitView.kt
package com.didichuxing.doraemonkit.kit.alignruler

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.WindowManager
import android.widget.FrameLayout
import com.blankj.utilcode.util.ActivityUtils
import com.blankj.utilcode.util.ConvertUtils
import com.blankj.utilcode.util.ScreenUtils
import com.didichuxing.doraemonkit.R
import com.didichuxing.doraemonkit.kit.alignruler.AlignRulerMarkerDokitView.OnAlignRulerMarkerPositionChangeListener
import com.didichuxing.doraemonkit.kit.core.AbsDokitView
import com.didichuxing.doraemonkit.kit.core.DokitViewLayoutParams
import com.didichuxing.doraemonkit.kit.core.DokitViewManager

/**
 * @author: xuchun
 * @time: 2020/6/4 - 10:47
 * @desc: 对齐标尺线View
 */
class AlignRulerLineDokitView : AbsDokitView(), OnAlignRulerMarkerPositionChangeListener {
    private var mMarker: AlignRulerMarkerDokitView? = null
    private var mAlignInfoView: AlignLineView? = null
    override fun onCreate(context: Context) {}
    override fun onDestroy() {
        super.onDestroy()
        mMarker?.removePositionChangeListener(this)
    }

    override fun onCreateView(context: Context, rootView: FrameLayout): View {
        return LayoutInflater.from(context).inflate(R.layout.dk_float_align_ruler_line, rootView, false)
    }

    override fun initDokitViewLayoutParams(params: DokitViewLayoutParams?) {
        params?.let {
            it.flags = WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE or WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE
            it.height = DokitViewLayoutParams.MATCH_PARENT
            it.width = DokitViewLayoutParams.MATCH_PARENT
        }

    }

    override fun onViewCreated(view: FrameLayout) {
        postDelayed(100, Runnable {
            mMarker = DokitViewManager.instance.getDokitView(ActivityUtils.getTopActivity(), AlignRulerMarkerDokitView::class.java.simpleName) as AlignRulerMarkerDokitView?
            mMarker?.addPositionChangeListener(this@AlignRulerLineDokitView)
        })
        setDokitViewNotResponseTouchEvent(rootView)
        mAlignInfoView = findViewById(R.id.info_view)
    }

    override fun onPositionChanged(x: Int, y: Int) {
        /**
         * 限制边界
         */
        var positionX = x
        var positionY = y
        if (!isNormalMode) {
            val iconSize = ConvertUtils.dp2px(30f)
            if (positionY <= iconSize) {
                positionY = iconSize
            }
            if (ScreenUtils.isPortrait()) {
                if (positionY >= screenLongSideLength - iconSize) {
                    positionY = screenLongSideLength - iconSize
                }
            } else {
                if (positionY >= screenShortSideLength - iconSize) {
                    positionY = screenShortSideLength - iconSize
                }
            }
            if (positionX <= iconSize) {
                positionX = iconSize
            }
            if (ScreenUtils.isPortrait()) {
                if (positionX >= screenShortSideLength - iconSize) {
                    positionX = screenShortSideLength - iconSize
                }
            } else {
                if (positionX >= screenLongSideLength - iconSize) {
                    positionX = screenLongSideLength - iconSize
                }
            }
        }
        mAlignInfoView?.showInfo(positionX, positionY)
    }

    override fun canDrag(): Boolean {
        return false
    }

    override fun restrictBorderline(): Boolean {
        return true
    }
}